docs/design/variadics.md

```diff
@@ -0,0 +1,658 @@
+compile-time loop over sequences called "packs". Packs are initialized and
+referred to using "pack bindings", which are marked with the `each` keyword at
+the point of declaration and the point of use.
+
+The syntax and behavior of a pack expansion depends on its context, and in some
+cases by a keyword following the `...`:
+
+-   In a tuple literal expression (such as a function call argument list), `...`
+    iteratively evaluates its operand expression, and treats the values as
+    successive elements of the tuple.
+-   `...and` and `...or` iteratively evaluate a boolean expression, combining
+    the values using `and` and `or`, and ending the loop early if the underlying
+    operator short-circuits.
+-   In a statement context, `...` iteratively executes a statement.
+-   In a tuple literal pattern (such as a function parameter list), `...`
+    iteratively matches the elements of the scrutinee tuple. In conjunction with
+    pack bindings, this enables functions to take an arbitrary number of
+    arguments.
+
+This example illustrates many of the key concepts:
+
+```carbon
+// Takes an arbitrary number of vectors with arbitrary element types, and
+// returns a vector of tuples where the i'th element of the vector is
+// a tuple of the i'th elements of the input vectors.
+fn Zip[... each ElementType:! type]
+      (... each vector: Vector(each ElementType))
+      -> Vector((... each ElementType)) {
+  ... var each iter: auto = each vector.Begin();
+  var result: Vector((... each ElementType));
+  while (...and each iter != each vector.End()) {
+    result.push_back((... each iter));
+    ... each iter++;
+  }
+  return result;
+}
+```
+
+A _pack_ is a sequence of a fixed number of values called "elements", which may
+be of different types. Packs are very similar to tuple values in many ways, but
+they are not first-class values -- in particular, packs do not have types, and
+no expression evaluates to a pack. The _arity_ of a pack is a compile-time value
+representing the number of values in the sequence.
+
+A pattern of the form "`...` _subpattern_" is called a _pack expansion pattern_.
+It can only appear as part of a tuple pattern (or an implicit parameter list),
+and it matches a sequence of tuple elements if each element matches
+_subpattern_. Since _subpattern_ will be matched against multiple scrutinees (or
+none) in a single pattern-matching operation, it cannot contain ordinary binding
+patterns. However, it can contain _pack binding patterns_, which are binding
+patterns that begin with `each`, such as `each ElementType:! type`. A pack
+binding pattern can match an arbitrarily large number of times, and binds to a
+pack consisting of all the matched values. A usage of a pack binding always has
+the form "`each` _identifier_".
+
+By default, a pack binding pattern can match any number of times, but `each` can
+optionally be followed by "`(>=` _integer-literal_ `)`", which constrains the
```

Comment:
While I was reading your nice proposal, I was thinking about `(>=` lakes the left operand. Is it possible to put `>=` next to `...`?

```cpp
...>=1 each param
```

In this way, it doesn't require parentheses, and the left operand of `>=` is `...`. Thanks.

Reply:
Yes, that's possible, and arguably it would be more accurate, because it actually does constrain the whole expansion. The main reason I attached it to `each` is that in that position, there's no ambiguity: `each` can only be followed by an identifier or an arity constraint, which are easy to distinguish. With `...` it's more subtle, because `...` can be followed by many things, including some operators. It can't otherwise be followed by `>=`, so it's not formally ambiguous, but it seems like it could easily lead to confusion for human readers, who might expect `...>=` to be related to `>=` in the same way that `...and` is related to `and`. Also, I'm already kind of uncomfortable with how we've given `...` so many syntactic variants (like `...and`, `...or`, and `...expand`), so I'm particularly reluctant to create yet another.

If we did attach it to the `...`, I think I'd want to keep the parentheses, in order to emphasize that this is nothing like `...and` and that the `1` is not part of the expansion body.